File: frontend/src/types/signalLinks.ts
import type { Direction, IndicatorType, Timeframe } from "./signal";

export type CloseOperator = ">=" | "<=";

export type SignalSnapshotInput = {
  indicator: IndicatorType;
  timeframe: Timeframe;
  value: number;
  direction: Direction;
};

export type CloseConditionInput = {
  operator: CloseOperator;
  targetValue: number;
};

export type LinkPositionSignalRequest = {
  symbol: string;
  signal: SignalSnapshotInput;
  closeCondition: CloseConditionInput | null;
};

export type LinkOrderSignalRequest = {
  symbol: string;
  exchangeOrderId: string;
  signal: SignalSnapshotInput;
  closeCondition: CloseConditionInput | null;
};

export type SignalSnapshotResponse = {
  id: number;
  indicator: IndicatorType;
  timeframe: Timeframe;
  value: number;
  direction: Direction;
};

export type CloseConditionResponse = {
  id: number;
  operator: CloseOperator;
  targetValue: number;
};

export type PositionSignalLinkResponse = {
  id: number;
  symbol: string;
  signal: SignalSnapshotResponse;
  closeCondition: CloseConditionResponse;
  createdAt: string;
};

export type OrderSignalLinkResponse = {
  id: number;
  symbol: string;
  exchangeOrderId: string;
  signal: SignalSnapshotResponse;
  closeCondition: CloseConditionResponse;
  createdAt: string;
};

export type CurrentIndicatorValue = {
  symbol: string;
  indicator: IndicatorType;
  value: number;
};

File: backend/schemas/signal_links.py
"""
Pydantic-схемы API для привязки сигналов к позициям и ордерам.

Используются в backend.api.signal_links для:
- валидации входящих запросов;
- формирования ответов API;
"""

from datetime import datetime

from backend.schemas.base import ApiModel


class SignalSnapshotInput(ApiModel):
    indicator: str
    timeframe: str
    value: float
    direction: str


class CloseConditionInput(ApiModel):
    operator: str
    target_value: float


class LinkPositionSignalRequest(ApiModel):
    """Запрос на привязку сигнала к открытой позиции."""
    symbol: str
    signal: SignalSnapshotInput
    close_condition: CloseConditionInput | None = None


class LinkOrderSignalRequest(ApiModel):
    """Запрос на привязку сигнала к открытому ордеру."""
    symbol: str
    exchange_order_id: str
    signal: SignalSnapshotInput
    close_condition: CloseConditionInput | None = None


class SignalSnapshotResponse(ApiModel):
    id: int
    indicator: str
    timeframe: str
    value: float
    direction: str


class CloseConditionResponse(ApiModel):
    id: int
    operator: str
    target_value: float


class PositionSignalLinkResponse(ApiModel):
    """Ответ с данными привязки сигнала к позиции."""
    id: int
    symbol: str
    signal: SignalSnapshotResponse
    close_condition: CloseConditionResponse | None = None
    created_at: datetime


class OrderSignalLinkResponse(ApiModel):
    """Ответ с данными привязки сигнала к ордеру."""
    id: int
    symbol: str
    exchange_order_id: str
    signal: SignalSnapshotResponse
    close_condition: CloseConditionResponse | None = None
    created_at: datetime